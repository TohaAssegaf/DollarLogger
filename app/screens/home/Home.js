import styles from './styles'
import HomeActionButton from '~/app/components/home/HomeActionButton'
import HomeWeekDetails from '~/app/components/home/HomeWeekDetails'
import { HEADER_BACKGROUND_COLOR, HEADER_TEXT_COLOR} from '~/app/config/colors'
import * as Routes from '~/app/config/Routes'
import * as DateUtils from '~/app/lib/DateUtils'
import * as PaymentUtils from '~/app/lib/PaymentUtils'
import moment from 'moment'
import React from 'react'
import { TouchableOpacity, View } from 'react-native'
import Swiper from 'react-native-swiper';
import Icon from 'react-native-vector-icons/Ionicons'
import { connect } from 'react-redux'

const ONE_MINUTE = 60000
const ONE_DAY = 86400000
const ONE_WEEK = ONE_DAY * 7
const THIS_WEEK = 'This week'
const LAST_WEEK = 'Last week'
const NEXT_WEEK = 'Next week'

class Home extends React.Component {
  static navigationOptions = ({ navigation }) => {
    const title = navigation.state.params && navigation.state.params.title
      ? navigation.state.params.title
      : THIS_WEEK
    return {
      title,
      headerStyle: {
        backgroundColor: HEADER_BACKGROUND_COLOR,
      },
      headerTintColor: HEADER_TEXT_COLOR,
      headerRight:
        <TouchableOpacity
            style={styles.settingsButton}
            onPress={() => navigation.navigate(Routes.SETTINGS)}>
          <Icon name="md-settings" size={25} color="white" />
        </TouchableOpacity>,
    }
  }

  componentDidMount() {
    this.intervalId = setInterval(
      () => {
        if (this.shouldSyncPayments()) {
          this.props.syncPayments()
        }
      },
      ONE_MINUTE)
  }

  /** Should sync payments if last successful fetch was before last failed fetch or push. */
  shouldSyncPayments() {
    return this.props.fetchSuccessTimestamp < this.props.fetchFailureTimestamp ||
      this.props.fetchSuccessTimestamp < this.props.pushFailureTimestamp
  }

  componentWillUnmount() {
    this.clearInterval(this.intervalId)
  }

  getHomeWeekDetailsList() {
    return this.props.paymentWeeks.map(
      (date, index) =>
        <HomeWeekDetails date={date} navigation={this.props.navigation} key={index}/>)
  }

  changeNavigationTitle(title: string) {
    this.props.navigation.setParams({ title })
  }

  onSwiperIndexChanged(index: number) {
    let title = moment(this.props.paymentWeeks[index]).format('MMMM Do YYYY')
    if (index === this.props.currentWeekIndex) {
      title = THIS_WEEK
    }
    if (index === this.props.currentWeekIndex - 1) {
      title = LAST_WEEK
    }
    if (index === this.props.currentWeekIndex + 1) {
      title = NEXT_WEEK
    }
    this.changeNavigationTitle(title)
  }

  render() {
    return (
      <View style={styles.screen}>
        <Swiper
            loop={false}
            index={this.props.currentWeekIndex}
            showsPagination={false}
            onIndexChanged={index => this.onSwiperIndexChanged(index)}>
          {this.getHomeWeekDetailsList()}
        </Swiper>
        <HomeActionButton
          addPaymentAction={() => this.props.navigation.navigate(Routes.ADD_PAYMENT)}
          historyAction={() => this.props.navigation.navigate(Routes.HISTORY)}
        />
      </View>
    );
  }
}

// TODO(renzobautista): Fix this to do an infinite scroll instead of doing 52 weeks before/after.
function getPaymentWeeks(paymentContributions: Array<PaymentContribution>): Array<Date> {
  const extraWeekCount: number = 52
  const currentMonday = DateUtils.getLastMonday(new Date())
  let dates = []
  for (let i = extraWeekCount; i > 0; i--) {
    dates.push(new Date(currentMonday.getTime() - ONE_WEEK * i))
  }
  dates.push(currentMonday)
  for (let i = 1; i <= extraWeekCount; i++) {
    dates.push(new Date(currentMonday.getTime() + ONE_WEEK * i))
  }
  return dates
}

const mapStateToProps = state => {
  const paymentContributions = PaymentUtils.getSortedPaymentContributions(state.payment.payments)
  const paymentWeeks = getPaymentWeeks(paymentContributions)
  const currentWeekIndex = paymentWeeks.map(
    date => date.getTime()).indexOf(DateUtils.getLastMonday(new Date()).getTime())
  return {
    paymentContributions,
    paymentWeeks,
    currentWeekIndex,
  }
}

const mapDispatchToProps = dispatch => {
  return {
    syncPayments: () => dispatch(actions.syncPayments())
  }
}


export default connect(mapStateToProps, mapDispatchToProps)(Home)
