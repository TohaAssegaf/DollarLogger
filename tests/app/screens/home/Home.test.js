import Home from '~/app/screens/home/Home'
import MockDate from 'mockdate'
import React from 'react';
import renderer from 'react-test-renderer';
import configureMockStore from 'redux-mock-store'
import { Provider } from 'react-redux'
import thunk from 'redux-thunk'

const middlewares = [thunk]
const mockStore = configureMockStore(middlewares)
jest.useFakeTimers()

it('renders correctly', () => {
  MockDate.set(new Date(2018, 3, 8))
  const store = mockStore({
    budget: { total: 25000 },
    payment: { payments: [
      {
        id: 1,
        total: 1,
        name: "Test payment",
        date: new Date(2018, 3, 8),
        paymentContributions: [
          {
            total: 1,
            date: new Date(2018, 3, 8),
            displayName: "Test payment",
            paymentId: 1,
          },
        ],
      }
    ]}})
  const rendered = renderer.create(
    <Provider store={store}><Home /></Provider>).toJSON();
  expect(rendered).toMatchSnapshot();
});

it('does not include previous week payments', () => {
  MockDate.set(new Date(2018, 3, 9))
  const store = mockStore({
    budget: { total: 25000 },
    payment: { payments: [
      {
        id: 1,
        total: 1,
        name: "Test payment",
        date: new Date(2018, 3, 8),
        paymentContributions: [
          {
            total: 1,
            date: new Date(2018, 3, 8),
            displayName: "Test payment",
            paymentId: 1,
          },
        ],
      }
    ]}})
  const rendered = renderer.create(
    <Provider store={store}><Home /></Provider>).toJSON();
  expect(rendered).toMatchSnapshot();
});
