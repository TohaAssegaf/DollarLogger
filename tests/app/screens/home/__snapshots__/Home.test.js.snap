// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View
  style={
    Object {
      "backgroundColor": "white",
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "borderBottomColor": "#222222",
        "borderBottomWidth": 0.5,
        "flexDirection": "row",
      }
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "flex": 0.5,
          "flexDirection": "column",
          "height": 100,
          "justifyContent": "center",
        }
      }
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "color": "black",
            "fontSize": 32,
          }
        }
      >
        $1,250.00
      </Text>
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
      >
        Remaining this week
      </Text>
    </View>
    <View
      style={
        Object {
          "alignItems": "center",
          "flex": 0.5,
          "flexDirection": "column",
          "height": 100,
          "justifyContent": "center",
        }
      }
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "color": "black",
            "fontSize": 32,
          }
        }
      >
        $1,250.00
      </Text>
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
      >
        Spent so far
      </Text>
    </View>
  </View>
  <View
    pointerEvents="box-none"
    style={
      Array [
        Array [
          Object {
            "backgroundColor": "transparent",
            "bottom": 0,
            "left": 0,
            "position": "absolute",
            "right": 0,
            "top": 0,
          },
          Object {
            "elevation": undefined,
            "justifyContent": "flex-end",
            "zIndex": undefined,
          },
        ],
        undefined,
      ]
    }
  >
    <View
      collapsable={undefined}
      pointerEvents="none"
      style={
        Object {
          "backgroundColor": "transparent",
          "bottom": 0,
          "elevation": undefined,
          "justifyContent": "flex-end",
          "left": 0,
          "opacity": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
          "zIndex": undefined,
        }
      }
    />
    <View
      pointerEvents="box-none"
      style={
        Array [
          Array [
            Object {
              "backgroundColor": "transparent",
              "bottom": 0,
              "left": 0,
              "position": "absolute",
              "right": 0,
              "top": 0,
            },
            Object {
              "elevation": undefined,
              "justifyContent": "flex-end",
              "zIndex": undefined,
            },
          ],
          Object {
            "alignItems": "flex-end",
          },
          Object {
            "paddingVertical": 30,
          },
        ]
      }
    >
      <View
        style={
          Array [
            Object {
              "marginHorizontal": 30,
              "zIndex": undefined,
            },
            Object {
              "elevation": 5,
              "shadowColor": "#000",
              "shadowOffset": Object {
                "height": 5,
                "width": 0,
              },
              "shadowOpacity": 0.35,
              "shadowRadius": 3,
            },
            undefined,
          ]
        }
      >
        <View
          accessibilityComponentType={undefined}
          accessibilityLabel={undefined}
          accessibilityTraits={undefined}
          accessible={true}
          collapsable={undefined}
          hasTVPreferredFocus={undefined}
          hitSlop={undefined}
          isTVSelectable={true}
          nativeID={undefined}
          onLayout={undefined}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
            }
          }
          testID={undefined}
          tvParallaxProperties={undefined}
        >
          <View
            collapsable={undefined}
            style={
              Object {
                "backgroundColor": "rgba(106, 27, 154, 1)",
                "borderRadius": 28,
                "height": 56,
                "width": 56,
              }
            }
          >
            <View
              collapsable={undefined}
              style={
                Object {
                  "alignItems": "center",
                  "borderRadius": 28,
                  "height": 56,
                  "justifyContent": "center",
                  "transform": Array [
                    Object {
                      "scale": 1,
                    },
                    Object {
                      "rotate": "0deg",
                    },
                  ],
                  "width": 56,
                }
              }
            >
              <Text
                accessible={true}
                allowFontScaling={true}
                collapsable={undefined}
                ellipsizeMode="tail"
                style={
                  Object {
                    "backgroundColor": "transparent",
                    "color": "rgba(255, 255, 255, 1)",
                    "fontSize": 24,
                    "marginTop": -4,
                  }
                }
              >
                +
              </Text>
            </View>
          </View>
        </View>
      </View>
    </View>
  </View>
</View>
`;
